googlemock/src/gmock-spec-builders.cc

```diff
@@ -551,19 +567,25 @@ class MockObjectRegistry {
     // "using ::std::cout;" doesn't work with Symbian's STLport, where cout is
     // a macro.
 
-    if (!GMOCK_FLAG(catch_leaked_mocks))
-      return;
-
     int leaked_count = 0;
     for (StateMap::const_iterator it = states_.begin(); it != states_.end();
          ++it) {
-      if (it->second.leakable)  // The user said it's fine to leak this object.
+      const MockObjectState& state = it->second;
+
+      if (!GMOCK_FLAG(catch_leaked_mocks)
+          || state.leakable)              // The user said it's fine to leak this object.
+      {
+        for (FunctionMockers::iterator mocker_it = state.function_mockers.begin();
+             mocker_it != state.function_mockers.end();
+             ++mocker_it) {
+           (*mocker_it)->SetRegistryDestructed();
```

Comment:
Hmmm...

Either:

1. Tell people using static-storage-lifetime-mocks, to compile googlemock as a shared lib. I've noticed that when setting up with `cmake -DBUILD_SHARED_LIBS=ON ..`, the global registry in the shared lib `gmock.so`, seems to outlive other statics. (But I'm not sure about the legalese, i.e. exact standards guarantees, etc. (Seems platform specific, so I'm not sure there can be any guarantees))  
OR
2. Put the global registry on the free store (heap) and just don't delete it. (Not sure that's a nice way to go? But see below: stuff like this is already in gmock in certain ifdefs...)
3. Possibly use a shared pointer that's held in mocks, to make lifetime guarantees (nice in C++11, but I suppose googlemock wants to be compatible with C98??) 

But note:

>Would life be easier if we just never allowed the global registry to die?

It's not just the registry that is involved. There are **other statics** that cause the same trouble. An example is the object pointed to by the pointer returned from `UnitTest::GetInstance()`: it is a static ([ref](https://github.com/google/googletest/blob/35fb11efbe1a2761ce923f49a9df1a430e5d16be/googletest/src/gtest.cc#L3986)), that (I think) dies too early in my [example 3](https://github.com/ajneu/global_mock/tree/master/3): Try it! -- the segfault is in the call-tree of `Expect` . 
(Funnily enough one finds... an ifdef Version that reads [as follows](https://github.com/google/googletest/blob/35fb11efbe1a2761ce923f49a9df1a430e5d16be/googletest/src/gtest.cc#L3983) - basically never allowing it to die.)

I'm not sure how many static objects in total, are involved in such lifetime issues...